File: 04 - Binary Trees/03 - Contains/BinaryTree.py
from __future__ import annotations


class Node:
    def __init__(self: Node, value: int) -> None:
        self.value: int = value
        self.left: Node | None = None
        self.right: Node | None = None


class BinaryTree:
    def __init__(self: BinaryTree) -> None:
        # We won't create an initial Node. The root will be None.
        self._root: Node | None = None

    @property
    def root(self: BinaryTree) -> Node | None:
        return self._root

    def insert(self: BinaryTree, new_value: int) -> bool:
        new_node: Node | None = Node(new_value)
        if self.root is None:
            self._root = new_node
            return True
        temp_node: Node | None = self.root
        while True:
            if new_node.value == temp_node.value:
                return False
            if new_node.value > temp_node.value:
                if temp_node.right is None:
                    temp_node.right = new_node
                    return True
                temp_node = temp_node.right
            else:
                if temp_node.left is None:
                    temp_node.left = new_node
                    return True
                temp_node = temp_node.left

    def contains(self: BinaryTree, value: int) -> bool:
        temp_node: Node | None = self.root
        # If the temp_node is None, it won't access the loop
        while temp_node is not None:
            if value == temp_node.value:
                return True
            if value > temp_node.value:
                temp_node = temp_node.right
            else:
                temp_node = temp_node.left
        return False


bt: BinaryTree = BinaryTree()

bt.insert(47)
bt.insert(21)
bt.insert(76)
bt.insert(18)
bt.insert(27)
bt.insert(52)
bt.insert(82)
bt.insert(15)
bt.insert(19)
bt.insert(25)
bt.insert(30)
bt.insert(49)
bt.insert(59)
bt.insert(77)
bt.insert(90)

#                     47
#          21                        76
#    18          27            52          82
# 15    19    25    30      49    59    77    90

print(bt.contains(47))
# True
print(bt.contains(21))
# True
print(bt.contains(25))
# True
print(bt.contains(76))
# True
print(bt.contains(52))
# True
print(bt.contains(77))
# True
print(bt.contains(0))
# False
print(bt.contains(100))
# False

File: 04 - Binary Trees/03 - Contains/BinaryTree.cs
public class BinaryTree
{
    public class Node
    {
        // The Node needs a value and pointers to Left and Right.
        public int Value { get; set; }
        public Node? Left { get; set; }
        public Node? Right { get; set; }
        public Node(int value) { this.Value = value; Left = null; Right = null; }
    }
    // A Binary Tree needs a "root".
    public Node? Root { get; private set; }
    public BinaryTree()
    {
        // We won't create an initial Node. The root will be 'null'.
        this.Root = null;
    }
    public bool Insert(int newValue)
    {
        // Edge cases: 
        // Tree is empty and the value already exists. 
        // We'll user a temp node to navigate through the tree.
        var newNode = new Node(newValue);
        if (Root == null)
        {
            Root = newNode;
            return true;
        }
        var tempNode = Root; // Temporary node to navigate through the tree
        while (true)
        {
            if (newNode.Value == tempNode.Value) // If the "new value" already exists.
                return false;
            if (newNode.Value > tempNode.Value)
            {
                if (tempNode.Right == null)
                {
                    tempNode.Right = newNode;
                    return true;
                }
                tempNode = tempNode.Right;
            }
            else
            {
                if (tempNode.Left == null)
                {
                    tempNode.Left = newNode;
                    return true;
                }
                tempNode = tempNode.Left;
            }
        }
    }
    public bool Contains(int value)
    {
        var tempNode = Root;
        while (tempNode != null) // If the tempNode is null, it won't access the loop
        {
            if (value == tempNode.Value)
                return true;
            if (value > tempNode.Value)
                tempNode = tempNode.Right;
            else
                tempNode = tempNode.Left;
        }
        return false;
    }
}